test/unit/spec/narrativeLoginSpec.js

```diff
@@ -1,19 +1,113 @@
+                responseText: JSON.stringify(responseObj)
+            });
+    }
+
+    describe('Test the kbaseNarrative module', () => {
+        beforeEach(() => {
+            setToken(FAKE_TOKEN);
+            jasmine.Ajax.install();
+        });
+
+        afterEach(() => {
+            jasmine.Ajax.uninstall();
+            clearToken();
+        });
+
+        it('Should instantiate and have expected functions', () => {
+            expect(Login).toBeDefined();
+            ['init', 'sessionInfo', 'getAuthToken', 'clearTokenCheckTimers'].forEach(item => {
+                expect(Login[item]).toBeDefined();
+            });
+        });
+
+        it('Should instantiate on a DOM node in the happy path', () => {
+            const $node = $('<div>'),
+                fakeUser = 'some_user',
+                fakeName = 'Some User',
+                tokenInfo = {
+                    expires: Date.now() + 10 * 60 * 60 * 24 * 1000,
+                    created: Date.now(),
+                    name: 'some_token',
+                    id: 'some_uuid',
+                    type: 'Login',
+                    user: 'some_user',
+                    cachefor: 500000
+                },
+                profileInfo = {
+                    created: Date.now() - 10000,
+                    lastlogin: Date.now(),
+                    display: fakeName,
+                    roles: [],
+                    customroles: [],
+                    user: fakeUser,
+                    local: false,
+                    email: `${fakeName}@kbase.us`,
+                    idents: []
+
+                };
+            // these should capture the happy path calls that Login.init should make
+            // copying some boilerplate from specs/api/authSpec
+            // TODO: move boilerplate to a mocks module
+
+            mockAuthRequest('token', tokenInfo, 200);
+            mockAuthRequest('me', profileInfo, 200);
+            return Login.init($node, true)  // true here means there's no kernel
+                .then(() => {
+                    const request = jasmine.Ajax.requests.mostRecent();
+                    expect(request.requestHeaders.Authorization).toEqual(FAKE_TOKEN);
+                    Login.clearTokenCheckTimers();
+                });
         });
 
-        it('Should instantiate on a DOM node', function() {
-            var $node = $("<div>");
-            Login.init($node);
+        it('Should throw an error when instantiating with a bad token', () => {
+            const $node = $('<div>');
+            mockAuthRequest('token', {error: {}}, 401);
+            mockAuthRequest('me', {error: {}}, 401);
+            return Login.init($node, true)  // true here means there's no kernel
+                .then(() => {
+                    const request = jasmine.Ajax.requests.mostRecent();
+                    expect(request.requestHeaders.Authorization).toEqual(FAKE_TOKEN);
+                    Login.clearTokenCheckTimers();
```

Comment:
It's difficult to tell from the tests that a DOM node has been instantiated in the first case and an error thrown in the second. What are the observable effects of the error being thrown (e.g. DOM structure changes?), and can they be tested here?

Reply:
This one was painful to work through. There's some asynchronous effects that happen in the narrativeLogin code at this point. The basic flow the `.init` function does is:
1. Talk to auth to validate the token and get the user profile.
2. Start up a timer to revalidate every second or so (without making an auth call, this timer also comes in to play if the user's computer went to sleep and wakes back up - if that's over some interval, THEN it makes an auth validation)
3. Use that info to initialize the UserMenu object that builds the little profile menu in the upper-right, with the gravatar and such.

The UserMenu then makes a call to the profile service to fetch gravatar info or user-set icon, and puts the image in. It's the part that does the work of manipulating the DOM. Since it does that in its own time, and the Promise isn't resolved by `NarrativeLogin.init`, there's likely no DOM changes that can be tested yet.

So there's 2 options here. Either leave as-is, or modify both the UserMenu code to have an explicit start function that returns a Promise, and pass that through the Login.init.

I'm not particularly against the code modification, in fact it's probably the right thing to do. I'm just trying to not do too many changes all at once here. But if you think it's a good idea, I'll put it in and test it. I think that module is (gasp) not unit tested anyway.

Reply:
Ugh, I think I just talked myself into it. I don't think it's a heavy change.